lib/fastMerge.js

```diff
@@ -1,99 +1,45 @@
-// Mostly copied from  https://medium.com/@lubaka.a/how-to-remove-lodash-performance-improvement-b306669ad0e1
+// Mostly copied from https://medium.com/@lubaka.a/how-to-remove-lodash-performance-improvement-b306669ad0e1
 
-/**
-* Is the object Mergeable
-*
-* @param val
-* @returns {*|boolean}
-*/
 function isMergeableObject(val) {
-  var nonNullObject = val && typeof val === 'object';
-  return nonNullObject &&
-    Object.prototype.toString.call(val) !== '[object RegExp]' &&
-    Object.prototype.toString.call(val) !== '[object Date]';
+    const nonNullObject = val && typeof val === 'object';
+    return nonNullObject
+    && Object.prototype.toString.call(val) !== '[object RegExp]'
+    && Object.prototype.toString.call(val) !== '[object Date]';
 }
-/**
-* Empty the Target
-*
-* @param val
-* @returns {*}
-*/
-function emptyTarget(val) {
-  return Array.isArray(val) ? [] : {}
-}
-/**
-* Clone if Necessary
-*
-* @param value
-* @param optionsArgument
-* @returns {*}
-*/
-function cloneIfNecessary(value, optionsArgument) {
-  var clone = optionsArgument && optionsArgument.clone === true;
-  return (clone && isMergeableObject(value)) ? merge(emptyTarget(value), value, optionsArgument) : value;
-}
-/**
-* Default Array Merge
-*
-* @param target
-* @param source
-* @param optionsArgument
-* @returns {*}
-*/
-let valArr;
-function defaultArrayMerge(target, source, optionsArgument) { 
-  var destination = target.slice();
-  for (let i = 0; i < source.length; ++i) {
-    valArr = source[i];
-    if (i >= destination.length) {
-      destination.push(cloneIfNecessary(valArr, optionsArgument));
-    } else if (typeof destination[i] === 'undefined') {
-      destination[i] = cloneIfNecessary(valArr, optionsArgument);
-    } else if (isMergeableObject(valArr)) {
-      destination[i] = merge(target[i], valArr, optionsArgument);
+
+function mergeObject(target, source) {
+    const destination = {};
+    if (isMergeableObject(target)) {
+        // eslint-disable-next-line rulesdir/prefer-underscore-method
+        const targetKeys = Object.keys(target);
+        for (let i = 0; i < targetKeys.length; ++i) {
+            const key = targetKeys[i];
+            destination[key] = target[key];
+        }
     }
-  }
-  return destination;
-}
-/**
-* Merge Object
-*
-* @param target
-* @param source
-* @param optionsArgument
-* @returns {{}}
-*/
-function mergeObject(target, source, optionsArgument) {
-  var destination = {};
-  if (isMergeableObject(target)) {
-    Object.keys(target).forEach(function (key) {
-      destination[key] = cloneIfNecessary(target[key], optionsArgument)
-    })
-  }
-  Object.keys(source).forEach(function (key) {
-    if (!isMergeableObject(source[key]) || !target[key]) {
-      destination[key] = cloneIfNecessary(source[key], optionsArgument)
-    } else {
-      destination[key] = merge(target[key], source[key], optionsArgument)
+    // eslint-disable-next-line rulesdir/prefer-underscore-method
```

Comment:
Why does the function has to be defined as `mergeObject = function (target, source)` instead of simply `function mergeObject(target, source)`?

Reply:
We have to do that otherwise we would use it before it's defined, since we have the following circular dependency `mergeObject -> fastMerge -> mergeObject`

Reply:
+1

Reply:
when 2 functions are on the same level, they are registered at the same time, it's just that eslint would flag it as a code style issue, because our code style is to define stuff before it's used - impossible to address in case of a circular dependency, but we can just disable the warning

The actual problem is with variables - you can't use a variable before it is defined, but you can use a function before it's defined, because functions are always registered before any code runs

Reply:
Thanks for clarifying @kidroca. In that case, I agree that we should declare it as `function mergeObject()` and disable the lint warning with a comment explaining why